### PR TITLE
[benchmark] BenchmarkDriver and BenchmarkDoctor

### PR DESCRIPTION
**Pull request by palimondo**
This is followup to #18124 that makes use of the direct memory measurements in `Benchmark_O` with the `--memory` option from the `Benchmark_Driver`. The PR also introduces thorough unit test coverage of the `Benchmark_Driver` and all newly added functionality. Summary of changes:

* Removes the legacy `Benchmark_Driver` [command `submit`](https://github.com/apple/swift/pull/18719/commits/343f284227ba8b151f8aee39f750a4ac172fb790) which is no longer used.
* Introduces [`test_utils.py`](https://github.com/apple/swift/blob/f38e6df914db58251002bf272cb03f3566d9906a/benchmark/scripts/test_utils.py) with homegrown unit testing helpers to support mocking.
* Creates [**`LogParser`**](https://github.com/apple/swift/blob/f38e6df914db58251002bf272cb03f3566d9906a/benchmark/scripts/compare_perf_tests.py#L280-L394) class for reading the output from `Benchark_O`, including `--verbose` mode.
* Creates [**`PerformanceTestsSamples`**](https://github.com/apple/swift/blob/f38e6df914db58251002bf272cb03f3566d9906a/benchmark/scripts/compare_perf_tests.py#L44-L186) class for computing thorough sample statistics, which can be used to exclude outliers.
* Creates [**`BenchmarkDriver`**](https://github.com/apple/swift/blob/f38e6df914db58251002bf272cb03f3566d9906a/benchmark/scripts/Benchmark_Driver#L32-L130) class for running benchmarks. This is invoked as a new `check` command to the `Benchmark_Driver`, with an optional `--verbose` parameter.
* Creates [**`BenchmarkDoctor`**](https://github.com/apple/swift/blob/f38e6df914db58251002bf272cb03f3566d9906a/benchmark/scripts/Benchmark_Driver#L159-L364) class for validating that benchmarks comply with the set of requirements (naming, runtime, memory) that ensure high quality of the Swift Benchmark Suite.
* `BenchmarkDriver` is now dependent on `compare_perf_tests.py`, so it is also [deployed to `swift-bin-dir` during build](https://github.com/apple/swift/pull/18719/files#diff-215555ba821abc0ddde0e78e1506e1e5R46).

The incremental rewrite of the `Benchmark_Driver`'s guts that adds full unit test coverage using the [StranglerApplication](https://www.martinfowler.com/bliki/StranglerApplication.html) refactoring pattern maintains the original functionality, with the exception of removing the bogus aggregate statistics from the last Totals row in the log. It now only reports the total number of executed benchmarks. The MAX_RSS column now reports the delta of memory used before and after running the benchmark, excluding the overhead of measurement infrastructure (i.e. the reported numbers are now lower by about 10MB).

I've structured the PR as a series of small, incremental changes with detailed descriptions in the commit messages, it is therefore best [**reviewed sequentially by individual commits**](https://github.com/apple/swift/pull/18719/commits/f29fef6b673596b004875a6e3822febbde565ce1).

### CODE REVIEW

**Comment by palimondo** [comment]
I've pushed fixes for minor formatting issues raised by python lint.

File: benchmark/scripts/Benchmark_Driver [hunk 1]
```diff
@@ -205,24 +212,101 @@ class BenchmarkDoctor(object):
                 'See http://bit.ly/UpperCamelCase')
 
     @staticmethod
-    def _name_is_at_most_40_chars_long(test):
-        if len(test) > 40:
-            BenchmarkDoctor.log_naming.warning(
-                "'%s' name is longer than 40 characters.", test)
+    def _name_is_at_most_40_chars_long(measurements):
```
Comment (eeckstein) [inline benchmark/scripts/Benchmark_Driver]: Not sure if this rule makes sense. If longer names cause troubles in reports, we should rather fix the reports

**Review by eeckstein — APPROVED** [review]
lgtm.
I didn't review in very detail, but I trust you that the changes will not break anything. It's great that you added all the test.

Regarding the test in Benchmark doctor which checks for naming convention: we can check for this but to be clear: just violating naming conventions does not make it worth renaming benchmarks and breaking history with this.

File: benchmark/scripts/Benchmark_Driver (hunk 1 above)
Comment (palimondo) [reply benchmark/scripts/Benchmark_Driver]: I have tried my best to fix the reports before, see the still open #9999. But there is no way to make it work on mobile. GitHub doesn’t give us much control over the CSS, which is the key.

If we follow your previous *rename-benchmark-when-runtime-changes-significantly* rule, we will be renaming most (all?) of the tests that violate the *max-length-40* naming rule while fixing the *runtime-under-2500-microseconds* violations anyway.

**Comment by palimondo** [comment]
I have tried my best to fix the reports before, see the still open #9999. But there is no way to make it work on mobile. GitHub doesn’t give us much control over the CSS, which is the key.

If we follow your previous precedent [*rename-benchmark-when-runtime-changes-significantly*](https://github.com/apple/swift/pull/18096#issuecomment-406758767) as a rule, we will be renaming most (all?) of the tests that violate the *max-length-40* naming rule while fixing the *runtime-under-2500-microseconds* rule violations anyway.

Could you ask @swift-ci to do python lint and benchmark, please?

**Comment by eeckstein** [comment]
> we will be renaming most (all?) of the tests that violate the max-length-40 naming rule while fixing the runtime-under-2500-microseconds rule violations anyway.

That's fine. But in general, I really don't want to mass-change/rename benchmarks which are _not_ noisy anyway, even if they have >2.5ms/iteration. It would cause a big disruption in our performance tracking.
We should only do that for problematic benchmarks.

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci Please Python lint

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci Please Python lint

**Comment by eeckstein** [comment]
ok, let's discuss it on the forum

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Pull request merged**